server/functions/framework/common_type.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package framework
+
+import (
+	"fmt"
+
+	"github.com/dolthub/go-mysql-server/sql"
+
+	pgtypes "github.com/dolthub/doltgresql/server/types"
+)
+
+// FindCommonType returns the common type that given types can convert to.
+// https://www.postgresql.org/docs/15/typeconv-union-case.html
+func FindCommonType(types []pgtypes.DoltgresTypeBaseID) (pgtypes.DoltgresTypeBaseID, bool) {
+	var candidateType = pgtypes.DoltgresTypeBaseID_Any
+	for _, typBaseID := range types {
+		if candidateType == pgtypes.DoltgresTypeBaseID_Any {
+			candidateType = typBaseID
+		} else if typBaseID == pgtypes.DoltgresTypeBaseID_Unknown {
+			continue
+		} else if candidateType != typBaseID {
+			if candidateType == pgtypes.DoltgresTypeBaseID_Unknown {
+				candidateType = typBaseID
+				continue
+			} else if candidateType.GetTypeCategory() != typBaseID.GetTypeCategory() {
+				return 0, false
+			} else if typCastFunction := GetImplicitCast(candidateType, typBaseID); typCastFunction == nil {
+				continue
+			}
+			candidateType = typBaseID
+			if candidateType.GetRepresentativeType().IsPreferredType() {
+				return candidateType, true
+			}
+		}
+	}
+	if candidateType == pgtypes.DoltgresTypeBaseID_Unknown {
+		return pgtypes.DoltgresTypeBaseID_Text, true
+	}
+	return candidateType, true
+}
```

Comment:
We can change this to something like the following:
1. Candidate starts out as the `unknown` type. Also, we'll add a fail flag that starts with `false`.
2. Loop over the given types the first time
2.1. Check if the iterated type matches the candidate.
2.2. If the types match, then we continue.
2.3. If the types do not match, then we check if the candidate is the `unknown` type.
2.4. If it is the `unknown` type, then we replace with the iterated type.
2.5. If it's not the `unknown` type, then we reset the candidate type to `unknown`, and set the fail flag to `true`.
3. If the fail flag is `false`, then we'll return the candidate type.
3.1. If the candidate type is `unknown`, then we'll return `text`.
4. Start another loop, this time checking the type categories. Fail if they belong to different categories (ignoring `unknown`).
5. Start a final loop, this time adhering to step 5 as closely as possible. This is similar to your current logic, however yours does not consider the implicit cast in both directions (whether to keep the candidate or replace it). In addition, we should continue to check for implicit conversions even after the preferred type is found, since that's what we would do in step 6, and we can verify that now while keeping the same logical flow, rather than erroring later.

This will seem like we're doing more work overall, but it's exactly the behavior outlined in the documentation. It's also easier to insert step 3 from the documentation later (which we'll ignore for now since we don't have domain types yet).

Reply:
The documentation says: `...If the candidate type can be implicitly converted to the other type, but not vice-versa, select the other type as the new candidate type.` 
I understood this as `candidate` implicitly casts to `other`, but `other` does not implicitly cast to `candidate`.

Reply:
I see what you mean. We're essentially selecting the first type that we come across, and one of the reasons that we switch is if we can't implicit cast to that first type. If we can't cast at all, then it's invalid, but if there's a cast going in the opposite direction, then we can replace it. Due to the requirement of all the types belonging in the same category, this effectively works like type promotion, since most casts are implicitly convertible to the larger or more general type, but they won't define implicit casts to the smaller types (just assignment or explicit).

This is further shown by step 6, since it will fail if there's no implicit conversion to the final candidate type. Also, we have to check the vice-versa statement, so we must check for the cast to know whether it's true.